### PR TITLE
Improve shadowing, support async sockets/callbacks in ZMQStream

### PR DESCRIPTION
**Pull request by minrk**
- can now create alternate socket APIs for existing sockets by passing one socket to another Socket constructor, with e.g. `zmq.asyncio.Socket(socket)`. Also works with `shadow`: `zmq.Socket.shadow(other_socket)` instead of needing `shadow(socket.underlying)`. This also preserves a reference to the shadowed object, avoiding close on garbage collection, which could happen if you tried to do e.g. `socket = zmq.asyncio.Socket.shadow(ctx.socket(zmq.PUSH).underlying)`
- add `socket_class` argument to `ctx.socket`, e.g. `ctx.socket(zmq.PUSH, socket_class=zmq.asyncio.Socket)` for easier creation of explicit socket types in a multi-socket-type situation
- Support async functions in ZMQStream callbacks
- Support async sockets in ZMQStream (by explicitly shadowing with a sync Socket, no matter what kind of Socket  it is)

The two seemingly unrelated changes are together because the async socket->sync socket shadow in zmqstream relies on the reference held when shadowing a socket object.

cc @blink1073 

In particular, this has consequences for anything that was _relying_ on the fact that `on_recv` callbacks were passed a Future. This _fixes_ that in that it will get the right argument now, but it means anything that had adopted and assumed the current (wrong) behavior will need to be updated. I'm not sure what the right way to 'fix' this is without breaking anything that assumed the broken behavior.

### CODE REVIEW

File: zmq/eventloop/zmqstream.py [hunk 1]
```diff
@@ -103,7 +125,16 @@ class ZMQStream:
     def __init__(
         self, socket: "zmq.Socket", io_loop: Optional["tornado.ioloop.IOLoop"] = None
     ):
+        if type(socket) is not zmq.Socket:
+            # shadow back to base zmq.Socket,
+            # otherwise callbacks like `on_recv` will get the wrong types.
+            # We know async sockets don't work,
+            # but other socket subclasses _may_.
+            # should we allow that?
+            # TODO: warn here?
```
[the remark this answers is not in the document]
Comment (minrk) [reply zmq/eventloop/zmqstream.py:134]: WDYT about warning in the async case? This is where behavior is changing, so a warning makes sense. But also now that shadowing is added, it works fine if the new behavior is what you expect, so a warning might be annoying!

[the remark this answers is not in the document]
Comment (blink1073) [reply zmq/eventloop/zmqstream.py:134]: Maybe instead error out if the original socket is async but the callback is not?  That way you're exposing the previously silent bad behavior.

[the remark this answers is not in the document]
Comment (blink1073) [reply zmq/eventloop/zmqstream.py:134]: Fair enough, that works

**Review by blink1073 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**